SCMOD-4128: Fix erroneous email detection

morvinshah4 opened the pull request:
https://jira.autonomy.com/browse/SCMOD-4128

morvinshah4 committed:
SCMOD-4128: Fixed erroneous email detection for emails tagged with blank headers

morvinshah4 committed:
Removing whitespace

morvinshah4 commented:
https://jira.autonomy.com/browse/SCMOD-4128

Anthony-McGreevy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java:
```diff
@@ -180,6 +180,24 @@ else if (line.contains(":*")) {
+                ((Element) emailElements.get(i-1).getContent().get(1)).setText(newBody);
+                emailElements.get(i).detach();
+            }
+        }
```
A lot of these variables can be final.

dermot-hardy replied:
Agreed - they should all be `final`.

dermot-hardy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java:
```diff
@@ -180,6 +180,24 @@ else if (line.contains(":*")) {
         }
         LOG.info("Markup of Headers and Body complete");
+
+        // Fix for SCMOD-4128 to untag and detach erroneous emails which have blank headers.
```
Don't mention the Jira number in the comments.  Just describe the behavior.  Describe for example why you are starting at index 1 rather than index 0.  A casual reader may think you've made a mistake with something unusual like that.

dermot-hardy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java:
```diff
@@ -180,6 +180,24 @@ else if (line.contains(":*")) {
+            if (header.getContentSize() == 0 && body.getContentSize() == 0)  {
+                emailElements.get(i).detach();
+            }
+            if (header.getContentSize() == 0) {
```
Is there a missing `else` here?

dermot-hardy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java:
```diff
@@ -180,6 +180,24 @@ else if (line.contains(":*")) {
+                emailElements.get(i).detach();
+            }
+            if (header.getContentSize() == 0) {
+                String currentBody = ((Element) emailElements.get(i).getContent().get(1)).getText();
```
Why is this not just `body.getText()`?

dermot-hardy reviewed (changes requested):
You need to some tests for this functionality.

morvinshah4 committed:
SCMOD-4128: Changes according to the review comments. Mainly variable declaration changed to final

morvinshah4 committed:
SCMOD-4128: Changed the loop mechanism to avoid exception

morvinshah4 committed:
SCMOD-4128: Moved the code to a seperate class as well as added a test case

Anthony-McGreevy commented:
There seems to be a lot of whitespace changes in the test class.

Commit:
Whitespace tidy-up

Commit:
Remove pointless import statement

Anthony-McGreevy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java:
```diff
@@ -0,0 +1,49 @@
+
+/**
+ * Created by Morvin Shah on 6/14/2018.
+ */
```
remove this comment

Anthony-McGreevy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java:
```diff
@@ -0,0 +1,49 @@
+ * Created by Morvin Shah on 6/14/2018.
+ */
+public final class EmailSquash {
+
```
maybe try and think of a better name that conveys what the class is being used for

morvinshah4 committed:
SCMOD-4128: Removed whitespaces

Anthony-McGreevy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java:
```diff
@@ -0,0 +1,49 @@
+
+    {
+        for (int i = emailElements.size() - 1; i > 0; i--) {
+            final Element header = (Element) emailElements.get(i).getContent().get(0);
```
add some comments to the method code and a javadoc to the method.

Anthony-McGreevy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/MarkupHeadersAndBody.java:
```diff
@@ -166,7 +167,7 @@ else if (line.contains(": ")) {
                     bodyIndex++;
                     addStandardisedHeader(nattyParser, headersElement, lines, line, ":\\*");
-                }else {
+                } else {
```
Remove whitespace changes from the commit. These are good to get in but they should be in their own commit, that way if we back out a feature we don't also lose good formatting changes.

Anthony-McGreevy commented on worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java:
```diff
@@ -278,6 +278,20 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
+    public void testEmailUntag() throws IOException, JDOMException
+    {
+       Document xmlDocument = TestUtility.readXmlFile("src/test/resources/xml/EmailUntag.xml");
+       MarkupHeadersAndBody.markUpHeadersAndBody(xmlDocument, emailHeaderMappings, condensedHeaderMultiLangMappings);
```
Make these variables final

Commit:
Create variables for repeatedly accessed propertties

Commit:
Updated not to repeat the email detach

Anthony-McGreevy commented on worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java:
```diff
@@ -278,6 +278,20 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
         assertTrue(TestUtility.compareHeaderElements(docForComparison, xmlDocument));
     }
 
+   @Test
```
Add a comment to say what this test is actually testing.

Commit:
Merge branch 'develop' into SCMOD-4128

Commit:
Removed changes unrelated to this Jira

Commit:
Corrected filename (Linux is case sensitive)

dermot-hardy commented on worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java:
```diff
@@ -278,6 +278,20 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
         assertTrue(TestUtility.compareHeaderElements(docForComparison, xmlDocument));
     }
 
+    @Test
```
All the other tests have comments describing what they do.  This one you've added doesn't.

dermot-hardy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java:
```diff
@@ -0,0 +1,50 @@
+    {
+    }
+
+    public static void untagFalseEmails(final List<Element> emailElements)
```
The class or method could do with some JavaDoc to describe what it is trying to achieve.

dermot-hardy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java:
```diff
@@ -0,0 +1,50 @@
+
+    public static void untagFalseEmails(final List<Element> emailElements)
+    {
+        for (int i = emailElements.size() - 1; i > 0; i--) {
```
Add a comment to explain why this is `i > 0` rather than `i >= 0`.

dermot-hardy reviewed (changes requested):
- So far as I can see the testcase that you've added doesn't actually test your code.  It doesn't contain any e-mails that don't have headers.
- On top of that you should have more that one testcase to test the different branches of your code and different placements of e-mails that need to be removed.

morvinshah4 committed:
SCMOD-4128: Added tests for multiple scenarios.

morvinshah4 committed:
SCMOD-4128: Added JavaDoc & comments on the code

morvinshah4 committed:
SCMOD-4128: Added licence header

Commit:
Tided up the comments

Commit:
Removed test files no longer used

Commit:
Changed variable names to match every other test function

dermot-hardy commented on worker-markup-core/src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBodyMarkedup.xml:
```diff
@@ -0,0 +1,133 @@
+    EDRM Enron Email Data Set has been produced in EML, PST and NSF format by ZL Technologies, Inc. This Data Set is licensed under a Creative Commons Attribution 3.0 United States License &lt;http://creativecommons.org/licenses/by/3.0/us/&gt; . To provide attribution, please cite to "ZL Technologies, Inc. (http://www.zlti.com)."
+    ***********
+
+</body></email><email><headers></headers><body></body></email></CONTENT><CUSTOMER_ID>morvinnew</CUSTOMER_ID><DATE_ARCHIVED>1527895585</DATE_ARCHIVED><DATE_CREATED>1527724821</DATE_CREATED><DATE_MODIFIED>1527724549</DATE_MODIFIED><DATE_PROCESSED>1527895585</DATE_PROCESSED><DOC_FORMAT_CODE>233</DOC_FORMAT_CODE><EXTRACTION_TYPE>Text</EXTRACTION_TYPE><FAMILY_ADDRESS_DISPLAY_FROM>"Dan J Hyvl" &lt;Dan  ...
```
If this test is passing then the logic must not work since I can clearly see an e-mail with no headers and no body:

    <email><headers></headers><body></body></email>

morvinshah4 replied:
Dermot I added the empty header and body after our call to test the other if condition

dermot-hardy replied:
This is the expectation file.  It is being tested against the output to confirm correct behavior.  To test the other `if` condition you'd need to adjust the input file, or better still add a separate test for it.

morvinshah4 committed:
SCMOD-4128: Changed the tests

Anthony-McGreevy commented on worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java:
```diff
@@ -284,15 +284,14 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
-        String docMarkedupValue = xmlDocument.getRootElement().getValue();
-        String docForComparisonValue = docForComparison.getRootElement().getValue();
+        String docMarkedupValue = expectedDocument.getRootElement().getValue();
+        String docForComparisonValue = actualDocument.getRootElement().getValue();
```
All of the variables in these test methods can be final.

morvinshah4 committed:
SCMOD-4128: Changed the variable names

morvinshah4 committed:
SCMOD-4128: Changed the variables to final

Anthony-McGreevy commented:
@morvinshah4 Is there a dev build of this where is can see these tests passing?

morvinshah4 commented:
http://sou-jenkins2.hpeswlab.net/job/CAFDataProcessing/view/Developer/job/CAFDataProcessing~worker-markup~SCMOD-4128~CI/

dermot-hardy reviewed (changes requested):
What we discussed earlier about the xml strings being compared hasn't been implemented.  When we talked about it you knew the correct functions to call but they are not being called here.

morvinshah4 committed:
SCMOD-4128: Changed the tests with the XMLOutputter method

dermot-hardy commented on worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java:
```diff
@@ -278,6 +279,38 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
+    {
+        final Document expectedDocument = TestUtility.readXmlFile("src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBody.xml");
+        MarkupHeadersAndBody.markUpHeadersAndBody(expectedDocument, emailHeaderMappings, condensedHeaderMultiLangMappings);
+        final Document actualDocument = TestUtility.readXmlFile("src/test/resources/xml/EmailUntagWithEmptyHeadersEmptyBodyMarkedup.xml");
```
This is the expected file - not the actual

dermot-hardy commented on worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java:
```diff
@@ -278,6 +279,38 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
+        final String docMarkedupValue = new XMLOutputter().outputString(actualDocument);
+        final String docForComparisonValue = new XMLOutputter().outputString(expectedDocument);
+
+        assertTrue(docForComparisonValue.equals(docMarkedupValue));
```
We should be using `assertEquals` for a case like this.

Commit:
Corrected the test code
- Actual and Expected variables were mixed up
- Use assertEquals rather than assertTrue

Commit:
Adjusted input file to remove one of the empty e-mails (leaving just one)

Commit:
Add checks that the values haven't changed

dermot-hardy reviewed (changes requested):
Looking at the testcases I can see that there may be `divider` elements between the `email` elements.  Looking at the code I can see that it doesn't take this into account.  The code looks OK if the invalid e-mail being squashed is preceded by a valid e-mail, but if it is preceded by a divider then it fails.

morvinshah4 committed:
SCMOD-4128: Now code takes divider as input. Also untags blank email after it

morvinshah4 committed:
SCMOD-4128: Added test case for Divider and empty email

dermot-hardy commented on worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java:
```diff
@@ -278,6 +279,65 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
+    /*
+     * Test for Untaging emails which have empty headers and empty body.
+     */
+   @Test
```
`@Test` annotation not aligned like the rest of them.

dermot-hardy commented on worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java:
```diff
@@ -278,6 +279,65 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
+    /*
+     * Test for Untaging emails which have empty headers and non empty body.
+     */
+   @Test
```
`@Test` annotation not aligned like the rest of them.

dermot-hardy commented on worker-markup-core/src/test/java/com/github/cafdataprocessing/worker/markup/core/MarkupOfHeadersAndBodyTest.java:
```diff
@@ -278,6 +279,65 @@ public void testFieldNameMappingOnMultipleEmailHeaderTag() throws IOException, J
+
+        assertEquals(expectedDocumentXml, actualDocumentXml);
+    }
     /**
```
Leave a blank line rather than running straight into the next function.

dermot-hardy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java:
```diff
@@ -0,0 +1,64 @@
+    public static void untagFalseEmails(final Element parentElement) {
+        final List<Element> emailElements = parentElement.getChildren();
+
+        // The loop starts with 1 because the we do not have header information of the parent email
```
This comment makes no sense - the loop doesn't start with 1.

dermot-hardy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java:
```diff
@@ -0,0 +1,64 @@
+        for (int i = emailElements.size() - 1; i > 0; i--) {
+            final Element previousElement = emailElements.get(i - 1);
+            final Element currentElement = emailElements.get(i);
+            INNERLOOP:
```
Why do we need to use a label here?  Also, what inner loop are you referring to here?  You appear to be labeling an `if` statement.

dermot-hardy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java:
```diff
@@ -0,0 +1,64 @@
+            final Element previousElement = emailElements.get(i - 1);
+            final Element currentElement = emailElements.get(i);
+            INNERLOOP:
+            if ((currentElement.getName().equals("email"))) {
```
There appear to be extra unnecessary sets of brackets in this line.

dermot-hardy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java:
```diff
@@ -0,0 +1,64 @@
+            if ((currentElement.getName().equals("email"))) {
+                final Element headersElement = (Element) currentElement.getContent().get(0);
+                final Element bodyElement = (Element) currentElement.getContent().get(1);
+                if ((previousElement.getName().equals("divider"))) {
```
There appear to be extra unnecessary sets of brackets in this line.

dermot-hardy commented on worker-markup-core/src/main/java/com/github/cafdataprocessing/worker/markup/core/EmailSquash.java:
```diff
@@ -0,0 +1,64 @@
+                final Element bodyElement = (Element) currentElement.getContent().get(1);
+                if ((previousElement.getName().equals("divider"))) {
+                    if (headersElement.getContentSize() == 0 && (bodyElement.getContentSize() == 0  ||
+                            bodyElement.getContent().get(0).getValue().equals(""))) {
```
- Add an additional check that there is only one content.  At the minute it appears to be an implicit assumption (although most likely a valid one).
- Use `.isEmpty()` rather than `.equals("")`

dermot-hardy commented on worker-markup-core/src/test/resources/xml/EmailUntagHavingDividerWithEmptyHeadersEmptyBody.xml:
```diff
@@ -0,0 +1,132 @@
+    limitations under the License.
+
+-->
+<root><ACCOUNTS>{"handles":["\"dan j hyvl\" &lt;dan j hyvl&gt;"],"identifier":"\"dan j hyvl\" &lt;dan j hyvl&gt;","sender":true,"identifier_type":"EML","display_name":"\"Dan J Hyvl\" &lt;Dan J Hyvl&gt;"}</ACCOUNTS><ACCOUNTS>{"handles":["CWPfister@cps-satx.com"],"identifier":"CWPfister@cps-satx.com","sender":false,"identifier_type":"EML","display_name":"Pfister Christian W."}</ACCOUNTS><ADDRESS_DI ...
```
The rest of the testcase files only have the `<CONTENT>` element and that is all that you require as well.  Drop all the redundant elements.

dermot-hardy replied:
This comment still hasn't been addressed.

morvinshah4 committed:
SCMOD-4128: Added additional check for only one content

morvinshah4 committed:
SCMOD-4128: Removed irrelevant fields

morvinshah4 committed:
SCMOD-4128: Indentation changes

morvinshah4 committed:
SCMOD-4128: Changed the comment

morvinshah4 committed:
SCMOD-4128: Removed irrelevant fields from test data

Commit:
Minor optimization to check the header count first

Commit:
Minor tidy-up: Whitespace changes only
